saleor/graphql/product/mutations/attributes.py

```diff
@@ -42,52 +37,148 @@ class Meta:
         model = models.Attribute
 
     @classmethod
-    def user_is_allowed(cls, user, input):
-        return user.has_perm('product.manage_products')
+    def clean_attribute_value_uniqueness(cls, values, errors, error_msg):
+        """Checks if all provided values are unique."""
+        if len(set(values)) != len(values):
+            cls.add_error(
+                errors,
+                cls.ATTRIBUTE_VALUES_FIELD,
+                error_msg)
+        return errors
 
     @classmethod
-    def clean_input(cls, info, instance, input, errors):
-        cleaned_input = super().clean_input(info, instance, input, errors)
-
-        values = cleaned_input.get('values', [])
-        names = [value['name'] for value in values]
-        if len(set(names)) != len(names):
-            cls.add_error(
-                errors, 'values', 'Duplicated attribute value names provided.')
+    def clean_attribute_values(cls, values, errors):
+        """Validates if a valid AttributeValue instance can be created using
+        the data provided.
+        """
         for value_data in values:
+            value_data['slug'] = slugify(value_data['name'])
             attribute_value = models.AttributeValue(**value_data)
             try:
                 attribute_value.full_clean()
             except ValidationError as validation_errors:
                 for field in validation_errors.message_dict:
-                    # Attribute instance is not created yet so we cannot add it
                     if field == 'attribute':
                         continue
                     for message in validation_errors.message_dict[field]:
-                        error_field = 'values:%(field)s' % {'field': field}
+                        error_field = '%(values_field)s:%(field)s' % {
+                            'values_field': cls.ATTRIBUTE_VALUES_FIELD,
+                            'field': field}
                         cls.add_error(errors, error_field, message)
+        return errors
+
+    @classmethod
+    def clean_input(cls, info, instance, input, errors):
```

Comment:
Could probably add a few comments/ refactor as this block is quite big?

Reply:
That's still heavy WIP